### PR TITLE
API server rate limiting

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

Trying to add some rate limiting to the API server. This will filter based on the IP present in the `X-Forwarded-For` header added by Traefik.

**Note**: this solution will not work on local deployments. All traffic is passing from the same IP.

**Note2**: the limit is very low. @pcrespov you might know better that me what to setup as limits.

## Checklist

- [ ] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [ ] Unit tests for the changes exist
- [ ] Runs in the swarm
- [ ] Documentation reflects the changes
- [ ] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

**Review by pcrespov — COMMENTED** [review]
Can you create a helper with a client that produces a given request rate? That will help us tuning the limit we wish to have. Could add that e.g. in ``pytest-simcore/helpers``

**Comment by GitHK** [comment]
> Can you create a helper with a client that produces a given request rate? That will help us tuning the limit we wish to have. Could add that e.g. in `pytest-simcore/helpers`

Yes I will try to have this in place.

**Review by sanderegg — APPROVED** [review]

**Comment by GitHK** [comment]
So I've added 2 utils to manually run when testing the rate limit of the service. Based on the parameters the test could run in seconds or in hours.

If you have suggestions on how to transform it into a test I am listening. Also the parameters could be picked up from the `docker-compose.yml` file to test all configured services automatically. I would do this only if were have a reliable "time boxed" test for bursts and steady rate testing for the rate limits.

@pcrespov @sanderegg

File: services/docker-compose.yml [hunk 1]
```diff
@@ -334,6 +334,15 @@ services:
...
           - node.role == manager
+      labels:
+        # for each service in the stack a new middlaware for rate limiting needs to be registered here
+        # requests = average / period this is how the limits are defined
+        - traefik.http.middlewares.ratelimit-${SWARM_STACK_NAME}_api-server.ratelimit.average=1
+        - traefik.http.middlewares.ratelimit-${SWARM_STACK_NAME}_api-server.ratelimit.period=1m
+        # a burst is computed over a period of 1 second
+        - traefik.http.middlewares.ratelimit-${SWARM_STACK_NAME}_api-server.ratelimit.burst=10
+        # X-Forwarded-For header extracts second IP from the right, count starts at one
+        - traefik.http.middlewares.ratelimit-${SWARM_STACK_NAME}_api-server.ratelimit.sourcecriterion.ipstrategy.depth=2
```
Comment (sanderegg) [inline services/docker-compose.yml:345]: for testing in devel or local mode. could you not add this with a different depth in docker-compose-local.yml?

**Review by sanderegg — APPROVED** [review]

File: services/docker-compose.yml (hunk 1 above)
Comment (GitHK) [reply services/docker-compose.yml:345]: It should be possible. *I think* these values will get tuned and more or less remain the same across all deployments.

File: packages/pytest-simcore/src/pytest_simcore/helpers/utils_rate_limit.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+from typing import List, Awaitable
+
+from aiohttp import ClientSession, ClientTimeout, ClientResponse
+
+log = logging.getLogger()
+
+
+def function_duration(func):
+    async def wrapper(*args, **kwargs):
```
Comment (pcrespov) [inline packages/pytest-simcore/src/pytest_simcore/helpers/utils_rate_limit.py:14]: MINOR: ``functools.wrap`` decorator?

**Review by pcrespov — APPROVED** [review]

**Pull request merged**